test/modules/TestResultCoordinator/report/CountingReportGenerator.cs

```diff
@@ -71,137 +59,94 @@ sealed class CountingReportGenerator : ITestResultReportGenerator
         /// <returns>Test Result Report.</returns>
         public async Task<ITestResultReport> CreateReportAsync()
         {
-            Logger.LogInformation($"Start to generate report by {nameof(CountingReportGenerator)} for Sources [{this.expectedSource}] and [{this.actualSource}]");
-
-            long lastLoadedKeyFromExpectStore = -1;
-            long lastLoadedKeyFromActualStore = -1;
             TestOperationResult lastLoadedResult = default(TestOperationResult);
-
-            var expectQueue = new Queue<TestOperationResult>();
-            var actualQueue = new Queue<TestOperationResult>();
             ulong totalExpectCount = 0;
             ulong totalMatchCount = 0;
             ulong totalDuplicateResultCount = 0;
             List<TestOperationResult> unmatchedResults = new List<TestOperationResult>();
-            ulong duplicatesFound = 0;
 
-            (lastLoadedKeyFromExpectStore, _, _) = await this.LoadBatchIntoQueueAsync(this.expectedSource, this.expectedStore, lastLoadedKeyFromExpectStore, expectQueue);
-            (lastLoadedKeyFromActualStore, lastLoadedResult, duplicatesFound) = await this.LoadBatchIntoQueueAsync(this.actualSource, this.actualStore, lastLoadedKeyFromActualStore, actualQueue, lastLoadedResult);
-            totalDuplicateResultCount += duplicatesFound;
+            bool hasExpectedResults = await this.ExpectedResults.MoveNextAsync();
+            TestOperationResult expectedResult = this.ExpectedResults.Current;
+            bool hasActualResults = await this.ActualResults.MoveNextAsync();
+            TestOperationResult actualResult = this.ActualResults.Current;
 
-            while (expectQueue.Count > 0 && actualQueue.Count > 0)
+            while (hasExpectedResults && hasActualResults)
             {
-                TestOperationResult expectedResult = expectQueue.Dequeue();
-                TestOperationResult actualResult = actualQueue.Peek();
+                this.ValidateStoredDataSource(this.ExpectedResults.Current, this.ExpectedSource);
+                this.ValidateStoredDataSource(this.ActualResults.Current, this.ActualSource);
+
+                while (hasActualResults && this.TestResultComparer.Matches(lastLoadedResult, actualResult))
+                {
+                    totalDuplicateResultCount++;
+                    hasActualResults = await this.ActualResults.MoveNextAsync();
+                    lastLoadedResult = actualResult;
+                    actualResult = this.ActualResults.Current;
+                }
 
                 totalExpectCount++;
 
-                if (this.testResultComparer.Matches(expectedResult, actualResult))
+                if (this.TestResultComparer.Matches(expectedResult, actualResult))
```

Comment:
I think you can directly use this.ExpectedTestResultSource.Current and this.ActualTestResultSource.Current here, right?